### PR TITLE
UI++

### PR DESCRIPTION
**Pull request by abesto**
Got familiar with the frontend code, made a bunch of improvements. Notably:
* Content now always fills complete window height (minor refactor around `Layout`)
* Sidebar doesn't start *under* the title
* Cleaned up the sidebar look a bit
* Syntax colors in function signatures
* Current module name is highlighted in the sidebar

There's more I want to do, but time for sleep, so here's a commit!

image

Note that generic type arguments will need an extra round of parsing.

I guess it'd be good to figure out what kind of design we're going for in general. My default is "look at rustdoc, do the same", but maybe you have something else in mind?

### CODE REVIEW

File: www/components/Function.tsx [hunk 1]
```diff
@@ -27,16 +34,32 @@ const Container = styled(Box)`
 const FuncDocs = styled(Documentation)`
   margin-left: 1em;
 `;
+const Keyword = styled.span`
+  font-weight: bold;
+  color: rgb(249, 38, 114);
```
Comment (abesto) [inline www/components/Function.tsx:39]: Ideally we'd load these colors from some theme, but we'll want to inject links etc., so just generating the code string and running it through a highlighter is not really an option.

File: www/components/Layout.tsx [hunk 2]
```diff
@@ -1,36 +1,40 @@
...
   title: string;
   children: ReactNode;
-}) {
+  sidebar?: ReactNode;
+}
+
+const Sidebar = styled.div`
+  overflow: auto;
+  background-color: #151515;
+  padding-left: ${(props) => props.theme.space.xxl};
```
Comment (abesto) [inline www/components/Layout.tsx:22]: This is the `styled-components` syntax for using values from the theme. WUI internally uses `@xstyled/styled-components`, and I tried to use that for a bit, but didn't get anywhere with it. If we can make that work, it's preferable; the syntax is much cleaner.

File: www/components/Layout.tsx [hunk 3]
```diff
@@ -1,36 +1,40 @@
...
+interface Props {
   title: string;
   children: ReactNode;
-}) {
+  sidebar?: ReactNode;
+}
+
+const Sidebar = styled.div`
+  overflow: auto;
+  background-color: #151515;
```
Comment (abesto) [inline www/components/Layout.tsx:21]: Whoops, ideally this should be another color from the theme (like below)

File: www/components/ModuleListSidebar.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import { Box } from "@welcome-ui/box";
+import { Text } from "@welcome-ui/text";
+import { Link } from "@welcome-ui/link";
+import styled from "styled-components";
+
+import { Pkg } from "../lib/docs";
+
+const ModuleLink = styled(Link)<{ active: boolean }>`
+  padding-left: ${(props) => props.theme.space.xs};
```
Comment (abesto) [inline www/components/ModuleListSidebar.tsx:11]: Ideally *all* spacing in CSS would be using values from the theme, for visual consistency.

File: www/components/Function.tsx [hunk 5]
```diff
@@ -12,7 +12,14 @@ const ParamWrapper = styled(Text)`
...
 function Param({ name, type }) {
-  const typeann = type === null ? "" : `: ${type}`;
+  const typeann =
+    type === null ? (
+      <></>
+    ) : (
+      <>
+        : <Ty>{type}</Ty>
+      </>
+    );
```
Comment (zsol) [inline www/components/Function.tsx:22]: okay, now with emojis!

File: www/components/Function.tsx [hunk 6]
```diff
@@ -27,16 +34,32 @@ const Container = styled(Box)`
...
-  const ret = func.returns === null ? "" : `-> ${func.returns}`;
+  const ret =
+    func.returns === null ? (
+      <></>
+    ) : (
+      <>
+        {" -> "}
+        <Ty>{func.returns}</Ty>
+      </>
+    );
```
Comment (zsol) [inline www/components/Function.tsx:56]: is this worth factoring out? `->` vs `:` can be a parameter

File: www/components/Layout.tsx [hunk 7]
```diff
@@ -1,36 +1,40 @@
...
-  min-height: 100vh;
-`;
-
 const theme = createTheme(darkTheme);
 
-export default function Layout({
-  title,
-  children,
-}: {
+interface Props {
```
Comment (zsol) [inline www/components/Layout.tsx:13]: teach me senpai, what this better?

File: www/components/Layout.tsx (hunk 2 above)
Comment (zsol) [reply www/components/Layout.tsx:22]: I started out with `@xstyled/styled-components` but it didn't play well with server side rendering and I couldn't make it work. Looks like next has better support for regular `styled-components`

File: www/components/ModuleListSidebar.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  padding-right: ${(props) => props.theme.space.md};
+  width: 100%;
+  ${(props) =>
+    props.active &&
+    `
+    background-color: ${props.theme.colors.light["500"]};
+    font-weight: bold;
+    color: ${props.theme.colors.dark["900"]};
+  `}
```
Comment (zsol) [inline www/components/ModuleListSidebar.tsx:20]: jesus... is this The Way?

File: www/components/ModuleListSidebar.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+`;
+
+interface ModuleListSidebarProps {
+  pkg: Pkg;
+  currentModule: string | undefined;
```
Comment (zsol) [inline www/components/ModuleListSidebar.tsx:25]: any reason for undefined vs null?

File: www/pages/index.tsx [hunk 10]
```diff
@@ -28,45 +31,46 @@ export default function Home({ packages }) {
           <ul>
             {packages.map(({ name, version }) => (
               <li key="{name}-{version}">
-                <Link href={'/' + name}><a>{name}</a></Link> ({version})
+                <Link href={"/" + name}>
+                  <a>{name}</a>
+                </Link>{" "}
+                ({version})
```
Comment (zsol) [inline www/pages/index.tsx:37]: not entirely convinced we need the version thing here. The deployed site will be large enough even with just the latest versions. 🤷

**Review by zsol — APPROVED** [review]

**Comment by zsol** [comment]
cc @anathien in case you wanna review :)

File: www/components/Function.tsx (hunk 6 above)
Comment (abesto) [reply www/components/Function.tsx:56]: Maybe? No strong opinion either way.

File: www/components/Layout.tsx (hunk 7 above)
Comment (abesto) [reply www/components/Layout.tsx:13]: Assuming s/what/why/.

It's mainly readability. To me the version with the interface inline is way harder to read; the same fields repeat with only `}: {` separating them, and it's hard for me to visually parse where one block starts and the other ends, especially when there are >3 props.

Another reason this is a good practice is it enables easily reusing the props specification if needed elsewhere. Like, if you want to wrap / extend the component, that's easier if you can refer to its props type explicitly.

File: www/components/ModuleListSidebar.tsx (hunk 8 above)
Comment (abesto) [reply www/components/ModuleListSidebar.tsx:20]: The alternative is one component with the shared styles, one component wrapping that with the active styles, and a third component switching between them. This felt cleaner (and it's *significantly* cleaner with `@xstyled` syntax)

File: www/components/ModuleListSidebar.tsx (hunk 9 above)
Comment (abesto) [reply www/components/ModuleListSidebar.tsx:25]: `currentModule` comes from the `symbol` variable in `[...pkg]`, which is `string | undefined`.

File: www/pages/index.tsx (hunk 10 above)
Comment (abesto) [reply www/pages/index.tsx:37]: I think the design should accommodate wanting to include version numbers at some point. Even if it doesn't, it's useful to show what version the docs are *for*.

File: www/components/Layout.tsx (hunk 7 above)
Comment (abesto) [reply www/components/Layout.tsx:13]: ... In an ideal world I'd want to write `function component({field: bool, otherField: string})` but TypeScript is not ideal that way :D

**Pull request merged**